interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp

```diff
@@ -122,79 +127,42 @@ namespace cling {
 
-      llvm::ArrayRef<Stmt*> stmtsRef(stmts.data(), stmts.size());
-      CompoundStmt* newCS = new (C) CompoundStmt(C, stmtsRef,
-                                                 CS->getLBracLoc(),
-                                                 CS->getRBracLoc());
-      return NodeContext(newCS);
-    }
-
-    NodeContext VisitIfStmt(IfStmt* If) {
-      // check the condition
-      NodeContext cond = Visit(If->getCond());
-      if (!cond.isSingleStmt())
-        If->setCond((clang::Expr*)cond.getStmt());
-
-      // No default constructor..
-      NodeContext result(If);
-      return result;
-    }
-
-    NodeContext VisitCastExpr(CastExpr* CE) {
-      NodeContext castExpr = Visit(CE->getSubExpr());
-      if (castExpr.isSingleStmt())
-        CE->setSubExpr((clang::Expr*)castExpr.getStmt());
+      if (!stmts.empty()) {
+        llvm::ArrayRef<Stmt*> stmtsRef(stmts.data(), stmts.size());
+        CS->setStmts(C, const_cast<Stmt**>(&stmtsRef.front()), stmts.size());
+      }
 
-      // No default constructor..
-      NodeContext result(CE);
-      return result;
+      return NodeContext(CS);
     }
 
     NodeContext VisitUnaryOperator(UnaryOperator* UnOp) {
       if (UnOp->getOpcode() == UO_Deref) {
-        UnOp->setSubExpr((clang::Expr*)SynthesizeCheck(UnOp->getLocStart(),
-                                 UnOp->getSubExpr()));
-      }
-
-      // No default constructor..
-      NodeContext result(UnOp);
-      return result;
-    }
-
-    NodeContext VisitBinaryOperator(BinaryOperator* BinOp) {
-      // Here we might get if(check) throw; binop rhs.
-      NodeContext rhs = Visit(BinOp->getRHS());
-      // Here we might get if(check) throw; binop lhs.
-      NodeContext lhs = Visit(BinOp->getLHS());
-
-      if (rhs.isSingleStmt()) {
-        // FIXME:we need to loop from 0 to n-1
-        BinOp->setRHS((clang::Expr*)rhs.getStmt());
-      }
-      if (lhs.isSingleStmt()) {
-        // FIXME:we need to loop from 0 to n-1
-        BinOp->setLHS((clang::Expr*)lhs.getStmt());
+        NodeContext newSubExpr = SynthesizeCheck(UnOp->getLocStart(),
+                                 UnOp->getSubExpr());
+        if (newSubExpr.isSingleStmt()) {
+          assert ((clang::Expr*)newSubExpr.getStmt()
+                  && "Visiting Stmt which is not an Expr!");
+          UnOp->setSubExpr((clang::Expr*)newSubExpr.getStmt());
+        }
       }
 
-      // No default constructor..
-      NodeContext result(BinOp);
-      return result;
+      return NodeContext(UnOp);
     }
 
     NodeContext VisitMemberExpr(MemberExpr* ME) {
       if (ME->isArrow()) {
         NodeContext newBase = SynthesizeCheck(ME->getLocStart(), ME->getBase());
         if (newBase.isSingleStmt())
+          assert ((clang::Expr*)newBase.getStmt()
+                  && "Visiting Stmt which is not an Expr!");
           ME->setBase((clang::Expr*)newBase.getStmt());
       }
 
-      NodeContext result(ME);
-      return result;
+      return NodeContext(ME);
```

Comment:
See above, we might want to recurse into `VisitStmt(ME)` here. Not sure whether we should do that before the call to `SynthesizeCheck()` or after.